File: packages/amplify_native_legacy_wrapper/android/src/main/kotlin/com/amazonaws/amplify/amplify_native_legacy_wrapper/AmplifyNativeLegacyWrapperPlugin.kt
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package com.amazonaws.amplify.amplify_native_legacy_wrapper

import android.content.Context
import com.amazonaws.amplify.amplify_native_legacy_wrapper.pigeons.LegacyNativePlugin
import com.amplifyframework.AmplifyException
import com.amplifyframework.auth.AuthException
import com.amplifyframework.auth.cognito.AWSCognitoAuthPlugin
import com.amplifyframework.kotlin.core.Amplify
import com.amplifyframework.core.AmplifyConfiguration
import io.flutter.embedding.engine.plugins.FlutterPlugin
import kotlinx.coroutines.runBlocking
import org.json.JSONObject


/** AmplifyNativeLegacyWrapperPlugin */
class AmplifyNativeLegacyWrapperPlugin: FlutterPlugin, LegacyNativePlugin {

  private lateinit var context: Context

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    context = flutterPluginBinding.applicationContext
    LegacyNativePlugin.setUp(flutterPluginBinding.binaryMessenger, this)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    LegacyNativePlugin.setUp(binding.binaryMessenger, null)
  }

  override fun configure(config: String, callback: (Result<Unit>) -> Unit) {
    try {
      val configuration = AmplifyConfiguration.builder(JSONObject(config))
        .devMenuEnabled(false)
        .build()
      Amplify.addPlugin(AWSCognitoAuthPlugin())
      Amplify.configure(configuration, context)
      callback(Result.success(Unit))
    } catch (e: AmplifyException) {
      callback(Result.failure(e))
    }
  }

  override fun signIn(username: String, password: String, callback: (Result<Unit>) -> Unit) {
    runBlocking {
      try {
        Amplify.Auth.signIn(username, password)
        callback(Result.success(Unit))
      } catch (error: AuthException) {
        callback(Result.failure(error))
      }
    }
  }

  override fun signOut(callback: (Result<Unit>) -> Unit) {
    runBlocking {
      try {
        Amplify.Auth.signOut()
        callback(Result.success(Unit))
      } catch (error: AuthException) {
        callback(Result.failure(error))
      }
    }
  }

  override fun rememberDevice(callback: (Result<Unit>) -> Unit) {
    runBlocking {
      try {
        Amplify.Auth.rememberDevice()
        callback(Result.success(Unit))
      } catch (error: AuthException) {
        callback(Result.failure(error))
      }
    }
  }

}

File: packages/amplify_native_legacy_wrapper/ios/Classes/AmplifyNativeLegacyWrapperPlugin.swift
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import Flutter
import AmplifyPlugins
import AWSPluginsCore
import Amplify
import UIKit

public class AmplifyNativeLegacyWrapperPlugin: NSObject, FlutterPlugin, LegacyNativePlugin {
    
    private let authPlugin: AWSCognitoAuthPlugin = AWSCognitoAuthPlugin()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let api = AmplifyNativeLegacyWrapperPlugin()
        let binaryMessenger = registrar.messenger()
        LegacyNativePluginSetup.setUp(binaryMessenger: binaryMessenger, api: api)
    }
    
    func configure(config: String, completion: @escaping (Result<Void, Error>) -> Void) {
        let decoder = JSONDecoder()
        do {
            let sourceData = config.data(using: .utf8)
            let decodedObject = try decoder.decode(JSONValue.self, from: sourceData!)
            let authConfig = decodedObject.value(at: "auth.plugins.awsCognitoAuthPlugin")
            try authPlugin.configure(using: authConfig)
            return completion(.success(()))
        } catch let error {
            return completion(.failure("Failed to configure Amplify: \(error)"))
        }
    }
    
    func signIn(username: String, password: String, completion: @escaping (Result<Void, Error>) -> Void) {
        _ = authPlugin.signIn(username: username, password: password, options: nil) { response in
            switch response {
            case .success(_):
                return completion(.success(()))
            case .failure(let signInError):
                return completion(.failure("Failed to sign the user in: \(signInError)"))
            }
        }
    }
    
    func signOut(completion: @escaping (Result<Void, Error>) -> Void) {
        _ = authPlugin.signOut(options: nil) { response in
            switch response {
            case .success(_):
                return completion(.success(()))
            case .failure(let signOutError):
                return completion(.failure("Failed to sign the user in: \(signOutError)"))
            }
        }
    }
    
    func rememberDevice(completion: @escaping (Result<Void, Error>) -> Void) {
        _ = authPlugin.rememberDevice(options: nil) { response in
            switch response {
            case .success(_):
                return completion(.success(()))
            case .failure(let error):
                return completion(.failure("Failed to remember device: \(error)"))
            }
        }
    }
    
}

extension String: Error {}


